qmi/core/transport_gpib_visa.py

```diff
@@ -0,0 +1,234 @@
+        """Read a specified number of bytes from the transport.
+
+        All bytes must belong to the same GPIB message.
+
+        This method blocks until the specified number of bytes are available,
+        then returns the received bytes. If timeout occurs, any partial read
+        data is discarded and QMI_TimeoutException is raised.
+
+        Parameters:
+            nbytes: Expected number of bytes to read.
+            timeout: Maximum time to wait in seconds (default: 60 seconds).
+
+        Returns:
+            Received bytes.
+
+        Raises:
+            ~qmi.core.exceptions.QMI_TimeoutException: If the timeout expires before the
+                requested number of bytes are available.
+            ~qmi.core.exceptions.QMI_EndOfInputException: If an end-of-message indicator
+                is received before the requested number of bytes is reached.
+        """
+        self._check_is_open()
+        nbuf = len(self._read_buffer)
+        while True:
+            if nbuf == nbytes:
+                # The requested number of bytes are already in the buffer. Return them immediately.
+                ret = self._read_buffer
+                self._read_buffer = bytes()
+                return ret
+
+            # GPIB requires a timeout
+            if timeout is None:
+                timeout = self._timeout
+
+            # Read buffer was of wrong length or is empty - read a new message from the instrument.
+            self._read_buffer += self._read_message(timeout)
+            nbuf = len(self._read_buffer)
+            if nbuf != nbytes:
+                _logger.debug("GPIB read buffer contained data %s" % self._read_buffer.decode())
+                self._read_buffer = bytes()
+                raise QMI_EndOfInputException(
+                    f"The read buffer did not contain expected bytes of data ({nbuf} != {nbytes}."
+                )
+
+    def read_until(self, message_terminator: bytes, timeout: Optional[float]) -> bytes:
+        """The specified message_terminator is ignored. Instead, the instrument must autonomously indicate the end
+        of the message according to the GPIB protocol.
+
+        As the message_terminator is not used, we forward simply the call to the `read_until_timeout` call.
+
+        Parameters:
+            message_terminator: This input is ignored.
+            timeout: Maximum time to wait (in seconds).
+
+        Returns:
+            Received bytes.
+        """
+        return self.read_until_timeout(0, timeout)
+
+    def read_until_timeout(self, nbytes: int, timeout: float) -> bytes:
+        """Read a single GPIB message from the instrument.
+
+        If there is already data in the buffer, return it.
+
+        If the timeout expires before the message is received, the read is
+        aborted and any data already received are discarded. In this
+        case QMI_TimeoutException is raised.
+
+        Parameters:
+            nbytes: This input is ignored.
```

Comment:
Correct, this should catch `QMI_TimeoutException` because that's what this call does. Not (re-)raise it... Fixed it, and fixed the similar mistake in `QMI_UsbTmcTransport` adn added unit-tests for it.